localstack/utils/docker.py

```diff
@@ -596,5 +803,559 @@ def mountable_tmp_file():
+            raise NoSuchContainer(container_name)
+        except APIError:
+            raise ContainerException()
+
+    def list_containers(self, filter: Union[List[str], str, None] = None, all=True) -> List[dict]:
+        if filter:
+            filter = [filter] if isinstance(filter, str) else filter
+            filter = dict([f.split("=") for f in filter])
+        LOG.debug("Listing containers with filters: %s", filter)
+        try:
+            container_list = self.client().containers.list(filters=filter, all=all)
+            return list(
+                map(
+                    lambda container: {
+                        "id": container.id,
+                        "image": container.image,
+                        "name": container.name,
+                        "status": container.status,
+                        "labels": container.labels,
+                    },
+                    container_list,
+                )
+            )
+        except APIError:
+            raise ContainerException()
+
+    def copy_into_container(
+        self, container_name: str, local_path: str, container_path: str
+    ) -> None:  # TODO behave like https://docs.docker.com/engine/reference/commandline/cp/
+        LOG.debug("Copying file %s into %s:%s", local_path, container_name, container_path)
+        try:
+            container = self.client().containers.get(container_name)
+            target_exists, target_isdir = self._container_path_info(container, container_path)
+            target_path = container_path if target_isdir else os.path.dirname(container_path)
+            with Util.tar_path(local_path, container_path, is_dir=target_isdir) as tar:
+                container.put_archive(target_path, tar)
+        except NotFound:
+            raise NoSuchContainer(container_name)
+        except APIError:
+            raise ContainerException()
+
+    def copy_from_container(
+        self,
+        container_name: str,
+        local_path: str,
+        container_path: str,
+    ) -> None:
+        LOG.debug("Copying file from %s:%s to %s", container_name, container_path, local_path)
+        try:
+            container = self.client().containers.get(container_name)
+            bits, _ = container.get_archive(container_path)
+            Util.untar_to_path(bits, local_path)
+        except NotFound:
+            raise NoSuchContainer(container_name)
+        except APIError:
+            raise ContainerException()
+
+    def pull_image(self, docker_image: str) -> None:
+        LOG.debug("Pulling image: %s", docker_image)
+        # some path in the docker image string indicates a custom repository
+        path_split = docker_image.rpartition("/")
+        image_split = path_split[2].partition(":")
+        repository = f"{path_split[0]}{path_split[1]}{image_split[0]}"
+        tag = image_split[2]
+        try:
+            LOG.debug("Repository: %s Tag: %s", repository, tag)
+            self.client().images.pull(repository, tag)
```

Comment:
I'm assuming the Docker Python SDK can handle `""` as the value for `tag` (treating it as `tag=None`)? (haven't tested this on my end, but probably works.. just to double-check)

Reply:
https://docker-py.readthedocs.io/en/stable/images.html Says: "If tag is None or empty, it is set to latest." So it should work.

This also happens in the test_docker_image_names test, where the alpine image is pulled without tag, which should result in the tag being "" for the docker sdk pull command.